src/cryptonote_core/service_node_list.cpp

```diff
@@ -507,23 +544,24 @@ namespace service_nodes
     quorum_state& quorum_state = m_quorum_states.next_free_slot();
```

Comment:
You may be storing a quorum for a height that already exists in the case of reorgs. So you have to search your existing buffer and see if an entry for the height already exists.

This would be easier to read, faster to implement, and more efficient as a map.